### DISCUSSION TITLE
weakKeys

### DISCUSSION THREAD

**Discussion by Jiej009** [ben-manes/caffeine#523] [Q&A] [answered]
Since garbage collection depends only on identity equality, this causes the whole cache to use identity (==) equality to compare keys, instead of equals ( ) .

why can't use equals(). Multiple object corresponding one keys data.
I can't understand the need not to use equals,Hope to teach me.

**Comment by ben-manes** [comment, marked as answer]
This is a carry over from Guava, where @kevinb9n [explained](https://groups.google.com/g/guava-discuss/c/_i1HdkaLDRU/m/GTx5pbXs1wIJ) the rational. There can be confusions because when the key that created the entry is GC'd then the entry is discarded. If you expect to find the entry by another equivalent key then your application may not hold the key and will observe more cache misses. In these cases, it may make more sense to use `weakValues`.

The primary case where `weakKeys` with equality may make sense is for a weak interner. This is what Guava does in `Interners.weak()`. The explicit contract of this interface is to discard duplicates and reuse a single instance from the cache (Flyweight pattern). In this case Guava uses a package-private `keyEquivalence` feature. That is not exposed as there were no good justifications internally for other scenarios. When maintaining the quality of a large corpus like Google's, there is a hesitancy to expose advanced functionality when it is assuredly going to be misused in almost all use-cases.

#344 does propose offering pluggable equivalences in Caffeine. That will likely happen someday as we can't target a single company's code base and instead should more strongly favor not blocking developers from getting their work done. That feature hasn't been implemented yet and is not being actively worked on.
